src/workload.rs

```diff
@@ -512,59 +620,173 @@ impl WorkloadInformation {
         }
     }
 
-    pub async fn find_upstream(&self, addr: SocketAddr, hbone_port: u16) -> Option<Upstream> {
-        self.fetch_address(&addr).await;
+    pub async fn find_upstream(
+        &self,
+        network: &str,
+        addr: SocketAddr,
+        hbone_port: u16,
+    ) -> Option<Upstream> {
+        self.fetch_address(network_addr(network, addr.ip())).await;
         let wi = self.info.lock().unwrap();
-        wi.find_upstream(addr, hbone_port)
+        wi.find_upstream(network, addr, hbone_port)
+    }
+
+    pub async fn find_waypoint(&self, wl: Workload) -> Result<Option<Upstream>, WaypointError> {
+        if let Some(gw_address) = &wl.waypoint {
```

Comment:
nit:

```rust

let Some(gw_address) = &wl.waypoint else {
  return Ok(None)
}
```

can lead to less indent needed

Reply:
addressed in https://github.com/istio/ztunnel/pull/486/commits/7b747454057e6cb8ec04db2803bcb6ce8728a11e